RFC/src/RFC-0170_NetworkCommunicationProtocol.md

```diff
@@ -0,0 +1,320 @@
+- When a CN receives an encrypted message, the node MUST attempt to open the message.
+- When a CN receives an encrypted message that the node is unable to open, and the destination node ID is known then the CN MUST forward it to all connected peers that have node IDs that are closer to the destination.
+- When a CN receives an encrypted message that the node is unable to open and the destination node is unknown then the CN MUST forward the message to all connected peers.
+- A CN MUST have the ability to verify the content of unencrypted messages to limit the propagation of spam messages.
+- If an unencrypted message is received by the CN with a unspecified destination node ID, then the node MUST verify the content of the message and forward the message to all connected peers.
+- If an unencrypted message is received by the CN with an specified destination node ID, then the node MUST verify the content of the message and forward the message to all connected peers that have closer node IDs.
+- A CN MUST have the ability to select a set of peer connections from its routing table.
+- Connections with the selected set of peers MUST be maintained by the CN.
+- A CN MUST have a mechanism to construct encrypted and unencrypted joining requests, discovery requests or data messages.
+- A CN MUST construct and provide a list of peer addresses from its routing table that is similar to a requested node ID so that other CCs and CNs can extend their routing tables.
+- A CN MUST keep its routing table up to date by removing unreachable peer addresses and adding newly received addresses.
+- It MUST have a mechanism to determine if a node ID was obtained through registration or was derived from an identification public key.
+- A CN MUST calculate the similarity between different node IDs by calculating the Hamming distance between the bits of the two node ID numbers.
+
+#### Functionality Required of Consensus Clients
+
+- It MUST select a cryptographic key pair used for identification on the Tari Communication network.
+- It MUST have a mechanism to derive a node ID from the self-selected identification public key.
+- A CC must have the ability to construct a peer address that links its identification public key, node ID and an online communication address.
+- A new CC MUST broadcast a joining request with its peer address to the Tari communication network so that CNs with similar node IDs can add the peer address of the new CC to their routing tables.
+- A CC MAY request the peer addresses of CNs with similar node IDs from other CNs to extend their local routing table.
+- A CC MUST have a mechanism to construct encrypted and unencrypted joining and discovery requests.
+- A CC MUST maintain a small persistent routing table of Tari Communication network peers with which ad hoc connections can be established.
```

Comment:
```suggestion
- A CC SHOULD maintain a small persistent routing table of Tari Communication network peers with which ad hoc connections can be established.
```

There is no way this can be enforced, by not keeping a small routing table you are effectively only limiting your self, it might be advantages to keep a larger one

Reply:
You must have a small routing table(with minimum 1 reliable peer connection) otherwise you can't connect to the network. I agree with you that it is beneficial to have the big routing table, but it is only recommended and not a requirement.

Reply:
It does not need to, but it could still receive messages from other peers. eg just a basenode listening on the network and not actively forwarding and participating in the network.

Reply:
I do not agree, no peers in routing table equals no connection or listening on Tari network.

Reply:
@SWvheerden MUST, SHOULD etc. are specification imperatives, so it's allowed to have a MUST directive even if it can't be enforced. Any code not following that directive is simply not written according to the spec.

The test here is: Will a node work without a routing table? I imagine one could provide a single peer address as part of the executable argument list (i.e. you build a peer list, but it's not persistent) and the node would still work, so that does suggest SHOULD